src/Functions/array/arrayLevenshtein.cpp

```diff
@@ -0,0 +1,368 @@
+            {
+                res_values[row] = static_cast<UInt32>(std::max(m, n));
+                continue;
+            }
+            PODArrayWithStackMemory<size_t, 32> v0(n + 1);
+
+            iota(v0.data() + 1, n, size_t(1));
+
+            for (size_t j = 1; j <= m; ++j)
+            {
+                v0[0] = j;
+                size_t prev = j - 1;
+                for (size_t i = 1; i <= n; ++i)
+                {
+                    size_t old = v0[i];
+                    v0[i] = std::min(prev + (lhs[j - 1] != rhs[i - 1]),
+                                     std::min(v0[i - 1], v0[i]) + 1);
+                    prev = old;
+                }
+            }
+            res_values[row] = static_cast<UInt32>(v0[n]);
+        }
+        return res;
+    }
+
+    ColumnPtr weightedLevenshteinImpl(std::vector<const ColumnArray *> columns, bool similarity) const
+    {
+        for (size_t i = 0; i < 2; i++)
+        {
+            const ColumnArray * hs_column = columns[i];
+            const ColumnArray * weights_column = columns[i+2];
+            for (size_t row = 0; row < hs_column->size(); row++)
+            {
+                Array hs = (*hs_column)[row].safeGet<Array>();
+                Array weights = (*weights_column)[row].safeGet<Array>();
+                if (hs.size() != weights.size())
+                    throw Exception(
+                        ErrorCodes::SIZES_OF_ARRAYS_DONT_MATCH,
+                        "Arguments {} ({}, size {}) and {} ({}, size {}) of function {} must be arrays of the same size.",
+                        toString(i + 1),
+                        hs_column->getName(),
+                        hs.size(),
+                        toString(i + 3),
+                        weights_column->getName(),
+                        weights.size(),
+                        getName());
+            }
+        }
+        auto res = ColumnFloat64::create();
+        ColumnFloat64::Container & res_values = res->getData();
+        res_values.resize(columns[0]->size());
+        for (size_t row = 0; row < columns[0]->size(); row++)
+        {
+            // Levenshtein with sliding vectors and weighted elements
+            // https://www.codeproject.com/Articles/13525/Fast-memory-efficient-Levenshtein-algorithm
+            Array lhs = (*columns[0])[row].safeGet<Array>();
+            Array rhs = (*columns[1])[row].safeGet<Array>();
+
+            Array array_l_weights = (*columns[2])[row].safeGet<Array>();
+            std::vector<Float64> l_weights(array_l_weights.size());
+            for (size_t i = 0; i < array_l_weights.size(); i++)
+                l_weights[i] = array_l_weights[i].safeGet<Float64>();
+            Array array_r_weights = (*columns[3])[row].safeGet<Array>();
+            std::vector<Float64> r_weights(array_r_weights.size());
+            for (size_t j = 0; j < array_r_weights.size(); j++)
+                r_weights[j] = array_r_weights[j].safeGet<Float64>();
```

Comment:
I am unsure if it's memory efficient or if it's better to use `safeGet` per element access. Maybe the latest commit is unnecessary.

Reply:
Found how to make it with lambda, and improved some other things